tools/node.d.ts

```diff
@@ -0,0 +1,207 @@
+    properties?: boolean;
+    pure_funcs?: string[];
+    pure_getters?: boolean | 'strict';
+    reduce_funcs?: boolean;
+    reduce_vars?: boolean;
+    sequences?: boolean;
+    side_effects?: boolean;
+    switches?: boolean;
+    toplevel?: boolean;
+    top_retain?: boolean;
+    typeofs?: boolean;
+    unsafe?: boolean;
+    unsafe_arrows?: boolean;
+    unsafe_comps?: boolean;
+    unsafe_Function?: boolean;
+    unsafe_math?: boolean;
+    unsafe_methods?: boolean;
+    unsafe_proto?: boolean;
+    unsafe_regexp?: boolean;
+    unsafe_undefined?: boolean;
+    unused?: boolean;
+    warnings?: boolean;
+}
+
+export enum InlineFunctions {
+    Disabled = 0,
+    SimpleFunctions = 1,
+    WithArguments = 2,
+    WithArgumentsAndVariables = 3
+}
+export interface MangleOptions {
+    eval?: boolean;
+    keep_classnames?: boolean;
+    keep_fnames?: boolean;
+    module?: boolean;
+    properties?: boolean | ManglePropertiesOptions;
+    reserved?: string[];
+    safari10?: boolean;
+    toplevel?: boolean;
+}
+
+export interface ManglePropertiesOptions {
+    builtins?: boolean;
+    debug?: boolean;
+    keep_quoted?: boolean;
+    regex?: RegExp;
+    reserved?: string[];
+}
+
+export interface OutputOptions {
+    ascii_only?: boolean;
+    beautify?: boolean;
+    braces?: boolean;
+    comments?: boolean | 'all' | 'some' | RegExp;
+    ecma?: ECMA;
+    indent_level?: number;
+    indent_start?: boolean;
+    inline_script?: boolean;
+    is8?: boolean;
+    keep_quoted_props?: boolean;
+    max_line_len?: boolean;
+    preamble?: string;
+    quote_keys?: boolean;
+    quote_style?: OutputQuoteStyle;
+    safari10?: boolean;
+    semicolons?: boolean;
+    shebang?: boolean;
+    shorthand?: boolean;
+    source_map?: SourceMapOptions;
+    webkit?: boolean;
+    width?: number;
+    wrap_iife?: boolean;
+}
+
+export enum OutputQuoteStyle {
+    PreferDouble = 0,
+    AlwaysSingle = 1,
+    AlwaysDouble = 2,
+    AlwaysOriginal = 3
+}
+
+export interface MinifyOptions {
+    compress?: boolean | CompressOptions;
+    ecma?: ECMA;
+    ie8?: boolean;
+    keep_classnames?: boolean;
+    keep_fnames?: boolean;
+    mangle?: boolean | MangleOptions;
+    module?: boolean;
+    nameCache?: object;
+    output?: OutputOptions;
+    parse?: ParseOptions;
+    safari10?: boolean;
+    sourceMap?: boolean | SourceMapOptions;
+    toplevel?: boolean;
+    warnings?: boolean | 'verbose';
+}
+
+export interface MinifyOutput {
+    ast?: AST_Node;
+    code?: string;
+    error?: Error;
+    map?: string;
+    warnings?: string[];
+}
+
+export interface SourceMapOptions {
+    content?: RawSourceMap;
+    includeSources?: boolean;
+    filename?: string;
+    root?: string;
+    url?: string | 'inline';
+}
+
+declare function parse(text: string, options?: ParseOptions): string;
+
+declare class AST_Node {
+    constructor(props: any);
+    BASE?: AST_Node;
+    PROPS: string[];
+    SELF_PROPS: string[];
+    SUBCLASSES: AST_Node[];
```

Comment:
Add: `CTOR: function;`

Either remove: `BASE`, `PROPS`, `SELF_PROPS`, `SUBCLASSES`, `documentation`, `propdoc`, `warn` and `from_mozilla_ast` **or** declare them as `static` properties. They are properties of the `AST_Node` `function` itself or the `AST_Node.CTOR` `function`:

```
$ node
> var T = require("terser");
undefined
> var node = new T.AST_String({ value: "hello" });
undefined

> node.PROPS
undefined
> node.SELF_PROPS
undefined
> node.SUBCLASSES
undefined
> node.BASE.TYPE
TypeError: Cannot read property 'TYPE' of undefined

> node.CTOR.PROPS
[ 'value', 'quote', 'start', 'end' ]
> node.CTOR.SELF_PROPS
[ 'value', 'quote' ]
> node.CTOR.SUBCLASSES
[]
> node.CTOR.BASE.TYPE
'Constant'
```

Reply:
Please add this as it's useful for debugging ASTs:
```
print_to_string(options?: OutputOptions);
```